core/src/main/java/org/apache/calcite/rex/RexFieldAccess.java

```diff
@@ -57,14 +59,23 @@
   RexFieldAccess(
       RexNode expr,
       RelDataTypeField field) {
+    checkValid(expr, field);
     this.expr = expr;
     this.field = field;
     this.digest = expr + "." + field.getName();
-    assert expr.getType().getFieldList().get(field.getIndex()) == field;
   }
 
   //~ Methods ----------------------------------------------------------------
 
+  private static void checkValid(RexNode expr, RelDataTypeField field) {
+    RelDataType exprType = expr.getType();
+    int fieldsNum = exprType.getFieldList().size();
+    int fieldIdx = field.getIndex();
+    final String errMsg = "Field " + field + " does not exist for expression " + expr;
+    Preconditions.checkArgument(fieldIdx < fieldsNum, errMsg);
+    Preconditions.checkArgument(exprType.getFieldList().get(fieldIdx).equals(field), errMsg);
+  }
```

Comment:
I made this change because there are some problem with the new added test case in test env with concurrency, the `==` check is invalid.